go/vt/vttablet/tabletserver/txthrottler/tx_throttler.go

```diff
@@ -251,31 +308,57 @@ func (t *txThrottler) Close() {
 	log.Info("txThrottler: closed")
 }
 
-// Throttle should be called before a new transaction is started.
-// It returns true if the transaction should not proceed (the caller
-// should back off). Throttle requires that Open() was previously called
-// successfully.
-func (t *txThrottler) Throttle(priority int, workload string) (result bool) {
+// Throttle should be called before a new transaction is started. It
+// returns an error if the transaction should not proceed (the caller
+// should back off). Throttle requires that Open() was previously
+// called successfully.
+func (t *txThrottler) Throttle(plan *planbuilder.Plan, options *querypb.ExecuteOptions) (err error) {
 	if !t.config.EnableTxThrottler {
-		return false
+		return nil
 	}
 	if t.state == nil {
-		return false
+		return nil
+	}
+
+	// get priority from execute options, skip if the priority
+	// is equal to sqlparser.MinPriorityValue
+	planType := plan.PlanID.String()
+	workload := options.GetWorkloadName()
+	metricLabels := []string{planType, workload}
+	t.requestsTotal.Add(metricLabels, 1)
+	priority := t.getPriorityFromOptions(options)
+	if priority == sqlparser.MinPriorityValue {
+		return nil
 	}
 
-	// Throttle according to both what the throttler state says and the priority. Workloads with lower priority value
-	// are less likely to be throttled.
-	result = t.state.throttle() && rand.Intn(sqlparser.MaxPriorityValue) < priority
+	// check if any throttling is needed
+	throttleErr := t.state.throttle(plan)
+	if throttleErr == nil {
+		return nil
+	}
 
-	t.requestsTotal.Add(workload, 1)
-	if result {
-		t.requestsThrottled.Add(workload, 1)
+	// Throttle probabilistically according to both what the throttler state says and the priority.
+	// Workloads with lower priority value are less likely to be throttled. "Hard" pool usage
+	// errors will throttle regardless of priority.
+	metricLabels = []string{planType, throttleErr.Error(), workload}
+	switch throttleErr {
+	case ErrThrottledConnPoolUsageHard, ErrThrottledTxPoolUsageHard:
```

Comment:
Say we have a connection pool of a 100 conns, and a hard threshold of 80. Let's also say that we are right at the threshold. A new query comes in. We get a Hard error, and the query gets throttled no matter what. That means that the remaining 20 conns never get used, does it not? Should we review this?